src/lib/web3/ethereum.ts

```diff
@@ -0,0 +1,192 @@
+
+export type EIP1193Provider = EIP1193Events & {
+  request: EIP1193RequestFn
+}
+
+export type EthereumWallet = EIP1193Provider
+
+/**
+ * EIP-4361 message fields
+ */
+export type SiweMessage = {
+  /**
+   * The Ethereum address performing the signing.
+   */
+  address: Address
+  /**
+   * The [EIP-155](https://eips.ethereum.org/EIPS/eip-155) Chain ID to which the session is bound,
+   */
+  chainId: number
+  /**
+   * [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) authority that is requesting the signing.
+   */
+  domain: string
+  /**
+   * Time when the signed authentication message is no longer valid.
+   */
+  expirationTime?: Date | undefined
+  /**
+   * Time when the message was generated, typically the current time.
+   */
+  issuedAt?: Date | undefined
+  /**
+   * A random string typically chosen by the relying party and used to prevent replay attacks.
+   */
+  nonce?: string
+  /**
+   * Time when the signed authentication message will become valid.
+   */
+  notBefore?: Date | undefined
+  /**
+   * A system-specific identifier that may be used to uniquely refer to the sign-in request.
+   */
+  requestId?: string | undefined
+  /**
+   * A list of information or references to information the user wishes to have resolved as part of authentication by the relying party.
+   */
+  resources?: string[] | undefined
+  /**
+   * [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#section-3.1) URI scheme of the origin of the request.
+   */
+  scheme?: string | undefined
+  /**
+   * A human-readable ASCII assertion that the user will sign.
+   */
+  statement?: string | undefined
+  /**
+   * [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) URI referring to the resource that is the subject of the signing (as in the subject of a claim).
+   */
+  uri: string
+  /**
+   * The current version of the SIWE Message.
+   */
+  version: '1'
+}
+
+export type EthereumSignInInput = SiweMessage
+
+// Helper functions
+
+function isAddress(address: string): boolean {
+  return /^0x[a-fA-F0-9]{40}$/.test(address)
+}
+
+export function getAddress(address: string): Address {
+  if (!isAddress(address)) {
+    throw new Error(`@supabase/auth-js: Address "${address}" is invalid.`)
+  }
+  return address.toLowerCase() as Address
+}
+
+export function fromHex(hex: Hex): number {
+  return parseInt(hex, 16)
+}
+
+export function toHex(value: string): Hex {
+  const bytes = new TextEncoder().encode(value)
+  const hex = Array.from(bytes, (byte) => byte.toString(16).padStart(2, '0')).join('')
+  return ('0x' + hex) as Hex
+}
+
+/**
+ * Creates EIP-4361 formatted message.
+ */
+export function createSiweMessage(parameters: SiweMessage): string {
+  const {
+    chainId,
+    domain,
+    expirationTime,
+    issuedAt = new Date(),
+    nonce,
+    notBefore,
+    requestId,
+    resources,
+    scheme,
+    uri,
+    version,
+  } = parameters
+
+  // Validate fields
+  {
+    if (chainId !== Math.floor(chainId))
```

Comment:
That's an interesting equality check I've not seen in a long time. Any specific reason for it, just curious?